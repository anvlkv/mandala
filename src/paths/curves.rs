use cfg_if::cfg_if;

use crate::{GlVec, Point, VectorValuedFn};

/// Quadratic Bezier curve with one control point
#[derive(Debug, Clone, Copy)]
pub struct QuadraticCurve {
    pub start: Point,
    pub control: Point,
    pub end: Point,
}

impl Default for QuadraticCurve {
    fn default() -> Self {
        Self {
            start: GlVec::default().into(),
            control: GlVec::default().into(),
            end: GlVec::default().into(),
        }
    }
}

impl VectorValuedFn for QuadraticCurve {
    fn eval(&self, t: crate::Float) -> crate::Vector {
        crate::Vector {
            x: (1.0 - t).powi(2) * self.start.x
                + 2.0 * (1.0 - t) * t * self.control.x
                + t.powi(2) * self.end.x,
            y: (1.0 - t).powi(2) * self.start.y
                + 2.0 * (1.0 - t) * t * self.control.y
                + t.powi(2) * self.end.y,
            #[cfg(feature = "3d")]
            z: (1.0 - t).powi(2) * self.start.z
                + 2.0 * (1.0 - t) * t * self.control.z
                + t.powi(2) * self.end.z,
        }
    }

    fn length(&self) -> crate::Float {
        let mut length = 0.0;
        let num_segments = 100;
        for i in 0..num_segments {
            let t1 = i as crate::Float / num_segments as crate::Float;
            let t2 = (i + 1) as crate::Float / num_segments as crate::Float;
            let p1 = self.eval(t1);
            let p2 = self.eval(t2);
            length += (p2.x - p1.x).hypot(p2.y - p1.y);
            cfg_if! { if #[cfg(feature = "3d")] {
                length += (p2.z - p1.z).abs();
            }}
        }
        length
    }
}

/// Cubic Bezier curve with two control points
#[derive(Debug, Clone, Copy)]
pub struct CubicCurve {
    pub start: Point,
    pub control1: Point,
    pub control2: Point,
    pub end: Point,
}

impl Default for CubicCurve {
    fn default() -> Self {
        Self {
            start: GlVec::default().into(),
            control1: GlVec::default().into(),
            control2: GlVec::default().into(),
            end: GlVec::default().into(),
        }
    }
}

impl VectorValuedFn for CubicCurve {
    fn eval(&self, t: crate::Float) -> crate::Vector {
        crate::Vector {
            x: (1.0 - t).powi(3) * self.start.x
                + 3.0 * (1.0 - t).powi(2) * t * self.control1.x
                + 3.0 * (1.0 - t) * t.powi(2) * self.control2.x
                + t.powi(3) * self.end.x,
            y: (1.0 - t).powi(3) * self.start.y
                + 3.0 * (1.0 - t).powi(2) * t * self.control1.y
                + 3.0 * (1.0 - t) * t.powi(2) * self.control2.y
                + t.powi(3) * self.end.y,
            #[cfg(feature = "3d")]
            z: (1.0 - t).powi(3) * self.start.z
                + 3.0 * (1.0 - t).powi(2) * t * self.control1.z
                + 3.0 * (1.0 - t) * t.powi(2) * self.control2.z
                + t.powi(3) * self.end.z,
        }
    }

    fn length(&self) -> crate::Float {
        let mut length = 0.0;
        let num_segments = 100;
        for i in 0..num_segments {
            let t1 = i as crate::Float / num_segments as crate::Float;
            let t2 = (i + 1) as crate::Float / num_segments as crate::Float;
            let p1 = self.eval(t1);
            let p2 = self.eval(t2);
            length += (p2.x - p1.x).hypot(p2.y - p1.y);
            cfg_if! { if #[cfg(feature = "3d")] {
                length += (p2.z - p1.z).abs();
            }}
        }
        length
    }
}

#[cfg(test)]
mod curve_tests {
    use super::*;
    use crate::test_util::test_name;
    use insta::assert_debug_snapshot;

    #[test]
    fn test_quadratic_curve_eval() {
        let curve = QuadraticCurve {
            start: Point {
                x: 0.0,
                y: 0.0,
                #[cfg(feature = "3d")]
                z: 0.0,
            },
            control: Point {
                x: 1.0,
                y: 1.0,
                #[cfg(feature = "3d")]
                z: 1.0,
            },
            end: Point {
                x: 2.0,
                y: 2.0,
                #[cfg(feature = "3d")]
                z: 2.0,
            },
        };
        assert_debug_snapshot!(
            test_name("QuadraticCurve"),
            [
                curve.eval(0.0),
                curve.eval(0.25),
                curve.eval(0.5),
                curve.eval(0.75),
                curve.eval(1.0),
            ]
        );
    }

    #[test]
    fn test_cubic_curve_eval() {
        let curve = CubicCurve {
            start: Point {
                x: 0.0,
                y: 0.0,
                #[cfg(feature = "3d")]
                z: 0.0,
            },
            control1: Point {
                x: 1.0,
                y: 1.0,
                #[cfg(feature = "3d")]
                z: 1.0,
            },
            control2: Point {
                x: 2.0,
                y: 2.0,
                #[cfg(feature = "3d")]
                z: 2.0,
            },
            end: Point {
                x: 3.0,
                y: 3.0,
                #[cfg(feature = "3d")]
                z: 3.0,
            },
        };
        assert_debug_snapshot!(
            test_name("CubicCurve"),
            [
                curve.eval(0.0),
                curve.eval(0.25),
                curve.eval(0.5),
                curve.eval(0.75),
                curve.eval(1.0),
            ]
        );
    }
}
