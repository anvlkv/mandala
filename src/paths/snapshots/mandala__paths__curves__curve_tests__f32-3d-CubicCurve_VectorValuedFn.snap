---
source: src/paths/curves.rs
expression: "(eval_points, sample_points, derivative_points, normal_points)"
---
(
    [
        Vector3 {
            x: 0.0,
            y: 0.0,
            z: 0.0,
        },
        Vector3 {
            x: 0.29999998,
            y: 2.9729998,
            z: 0.29999998,
        },
        Vector3 {
            x: 0.6,
            y: 5.784001,
            z: 0.6,
        },
        Vector3 {
            x: 0.90000004,
            y: 8.271,
            z: 0.90000004,
        },
        Vector3 {
            x: 1.2,
            y: 10.272,
            z: 1.2,
        },
        Vector3 {
            x: 1.5,
            y: 11.625,
            z: 1.5,
        },
        Vector3 {
            x: 1.8,
            y: 12.167999,
            z: 1.8,
        },
        Vector3 {
            x: 2.1,
            y: 11.739,
            z: 2.1,
        },
        Vector3 {
            x: 2.4,
            y: 10.176001,
            z: 2.4,
        },
        Vector3 {
            x: 2.6999998,
            y: 7.317001,
            z: 2.6999998,
        },
        Vector3 {
            x: 3.0,
            y: 3.0,
            z: 3.0,
        },
    ],
    [
        Vector3 {
            x: 0.0,
            y: 0.0,
            z: 0.0,
        },
        Vector3 {
            x: 0.3333333,
            y: 3.2962964,
            z: 0.3333333,
        },
        Vector3 {
            x: 0.6666666,
            y: 6.3703704,
            z: 0.6666666,
        },
        Vector3 {
            x: 0.99999994,
            y: 8.999999,
            z: 0.99999994,
        },
        Vector3 {
            x: 1.3333334,
            y: 10.962963,
            z: 1.3333334,
        },
        Vector3 {
            x: 1.6666667,
            y: 12.037038,
            z: 1.6666667,
        },
        Vector3 {
            x: 2.0,
            y: 12.0,
            z: 2.0,
        },
        Vector3 {
            x: 2.3333333,
            y: 10.629628,
            z: 2.3333333,
        },
        Vector3 {
            x: 2.6666665,
            y: 7.703703,
            z: 2.6666665,
        },
        Vector3 {
            x: 3.0,
            y: 3.0,
            z: 3.0,
        },
    ],
    [
        Vector3 {
            x: 3.0,
            y: 30.0,
            z: 3.0,
        },
        Vector3 {
            x: 3.125,
            y: 29.0,
            z: 3.125,
        },
        Vector3 {
            x: 2.75,
            y: 24.0,
            z: 2.75,
        },
        Vector3 {
            x: 3.0,
            y: 20.0,
            z: 3.0,
        },
        Vector3 {
            x: 3.0,
            y: 24.0,
            z: 3.0,
        },
        Vector3 {
            x: 3.0,
            y: 8.0,
            z: 3.0,
        },
        Vector3 {
            x: 3.0,
            y: 4.0,
            z: 3.0,
        },
        Vector3 {
            x: 3.0,
            y: -8.0,
            z: 3.0,
        },
        Vector3 {
            x: 3.0,
            y: -20.0,
            z: 3.0,
        },
        Vector3 {
            x: 2.0,
            y: -40.0,
            z: 2.0,
        },
        Vector3 {
            x: 2.0,
            y: -52.0,
            z: 2.0,
        },
    ],
    [
        Vector3 {
            x: -0.08920646,
            y: 0.10793537,
            z: -0.99014753,
        },
        Vector3 {
            x: -0.09517407,
            y: 0.11678457,
            z: -0.9885865,
        },
        Vector3 {
            x: -0.1003061,
            y: 0.124601305,
            z: -0.9871237,
        },
        Vector3 {
            x: -0.12517338,
            y: 0.16551077,
            z: -0.978232,
        },
        Vector3 {
            x: -0.1079272,
            y: 0.13658237,
            z: -0.9847319,
        },
        Vector3 {
            x: -0.21984835,
            y: 0.4137377,
            z: -0.8834522,
        },
        Vector3 {
            x: -0.23304205,
            y: 0.6892773,
            z: -0.6859944,
        },
        Vector3 {
            x: 0.21984835,
            y: 0.4137377,
            z: 0.8834522,
        },
        Vector3 {
            x: 0.12517338,
            y: 0.16551077,
            z: 0.978232,
        },
        Vector3 {
            x: 0.047387753,
            y: 0.05224496,
            z: 0.9975093,
        },
        Vector3 {
            x: 0.036929805,
            y: 0.03982514,
            z: 0.998524,
        },
    ],
)
